### PR TITLE
[SYCL-MLIR] Use GlobalVarAddressSpace for globals

### PR DESCRIPTION
**Pull request by whitneywhtsang**
Signed-off-by: Tsang, Whitney <whitney.tsang@intel.com>

### CODE REVIEW

File: polygeist/tools/cgeist/Lib/CGExpr.cc [hunk 1]
```diff
@@ -1435,13 +1435,16 @@ ValueCategory MLIRScanner::VisitDeclRefExpr(DeclRefExpr *E) {
...
     bool isArray = gv.second;
     // TODO check reference
     if (isArray)
-      return ValueCategory(gv2, /*isReference*/ true);
+      return ValueCategory(V, /*isReference*/ true);
     else
-      return ValueCategory(gv2, /*isReference*/ true);
-    // return gv2;
+      return ValueCategory(V, /*isReference*/ true);
+    // return V;
```
Comment (Naghasan) [inline polygeist/tools/cgeist/Lib/CGExpr.cc:1447]: Do we need to keep the commented code ?

Also both branches do the same thing.

File: polygeist/tools/cgeist/Lib/clang-mlir.cc [hunk 2]
```diff
@@ -2510,7 +2510,13 @@ MLIRASTConsumer::GetOrCreateGlobal(const ValueDecl *FD, std::string prefix,
   }
 
   mlir::Type rt = getTypes().getMLIRType(FD->getType());
-  unsigned memspace = 0;
+  auto VD = dyn_cast<VarDecl>(FD);
```
Comment (victor-eds) [inline polygeist/tools/cgeist/Lib/clang-mlir.cc:2513]:
```suggestion
  auto *VD = dyn_cast<VarDecl>(FD);
```

**Review by victor-eds — CHANGES_REQUESTED** [review]
LGTM but for NIT.

File: polygeist/tools/cgeist/Lib/CGExpr.cc (hunk 1 above)
Comment (whitneywhtsang) [reply polygeist/tools/cgeist/Lib/CGExpr.cc:1447]: Not sure why the code was like that. wasn't sure if I should modify it in this PR. I can change it.

File: polygeist/tools/cgeist/Lib/clang-mlir.cc [hunk 3]
```diff
@@ -2510,7 +2510,13 @@ MLIRASTConsumer::GetOrCreateGlobal(const ValueDecl *FD, std::string prefix,
...
+  if (!VD)
+    FD->dump();
+  VD = VD->getCanonicalDecl();
+  unsigned memspace = VD ? CGM.getContext().getTargetAddressSpace(
+                               CGM.GetGlobalVarAddressSpace(VD))
+                         : 0;
```
[the remark this answers is not in the document]
Comment (etiotto) [reply polygeist/tools/cgeist/Lib/clang-mlir.cc]: I think we should use the same address space infrastructure as clang does.  That is, in CodeGenModule.cpp we have this function: `LangAS CodeGenModule::GetGlobalVarAddressSpace(const VarDecl *D)`. Can we use that one ?

[the remark this answers is not in the document]
Comment (etiotto) [reply polygeist/tools/cgeist/Lib/clang-mlir.cc]: Essentially, I think the more future-proof solution is to port `TargetInfo.cpp`.

[the remark this answers is not in the document]
Comment (whitneywhtsang) [reply polygeist/tools/cgeist/Lib/clang-mlir.cc]: I am using `GetGlobalVarAddressSpace`.

**Review by victor-eds — APPROVED** [review]

File: polygeist/tools/cgeist/Lib/clang-mlir.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (etiotto) [reply polygeist/tools/cgeist/Lib/clang-mlir.cc]: Perfect.

File: polygeist/tools/cgeist/Lib/clang-mlir.cc [hunk 4]
```diff
@@ -2510,7 +2510,13 @@ MLIRASTConsumer::GetOrCreateGlobal(const ValueDecl *FD, std::string prefix,
   }
 
   mlir::Type rt = getTypes().getMLIRType(FD->getType());
-  unsigned memspace = 0;
+  auto *VD = dyn_cast<VarDecl>(FD);
+  if (!VD)
+    FD->dump();
```
Comment (etiotto) [inline polygeist/tools/cgeist/Lib/clang-mlir.cc:2515]: naked trace, use LLVM_DEBUG pls

**Review by etiotto — APPROVED** [review]

**Pull request merged**